README.md

```diff
@@ -434,4 +627,440 @@ asyncio.run(main())
+            slf.waker = Some(Arc::new(WakerClosure {
+                aio_loop,
+                py_future,
+            }));
+            Ok(())
+        })?;
+
+        Ok(slf)
+    }
+}
+
+#[pyproto]
+impl PyIterProtocol for MyFuture {
+    // Now MyFuture has truly become a Future wrapper
+    fn __next__(mut py_slf: PyRefMut<Self>) -> IterNextOutput<PyRefMut<Self>, u32> {
+        // This is so we can borrow both slf.waker & slf.future mutably
+        let slf = &mut *py_slf;
+        let waker = slf.waker.as_mut().expect("future was not awaited");
+        // Thanks to `futures` we don't need to implement the Waker pseudo-trait
+        // manually, so this is all boiler-plate to pass the Waker to the Future.
+        let waker_ref = waker_ref(&waker);
+        let context = &mut Context::from_waker(&*waker_ref);
+        match slf.future.as_mut().poll(context) {
+            Poll::Pending => {
+                // In case the future needs to be put on hold multiple times,
+                // we need to set this back to true every time we're in a
+                // Pending state. Asyncio always sets it to false after
+                // reading it when we yield ourself.
+                slf._asyncio_future_blocking = true;
+                IterNextOutput::Yield(py_slf)
+            }
+            Poll::Ready(result) => IterNextOutput::Return(result),
+        }
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+struct WakerClosure {
+    aio_loop: PyObject,
+    py_future: Py<MyFuture>,
+}
+
+// WakerClosure now implements `futures::task::ArcWake` so it can be passed to a
+// Rust Future as a waker, like we did above.
+impl ArcWake for WakerClosure {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        let closure = (**arc_self).clone();
+        Python::with_gil(|py| {
+            arc_self
+                .aio_loop
+                .call_method1(py, "call_soon_threadsafe", (closure,))
+        })
+        .expect("exception thrown by the event loop (probably closed)");
+        // Again, as a reminder, we are only setting ourself up to be called by
+        // the event loop from the main thread so this child thread is released
+        // before the task is woken up. Remember how the TimerFuture still has
+        // its shared state locked when calling `.wake()`? This is where it
+        // would have become a problem if we tried to call the callbacks directly.
+        // Of course, here, we could fix the issue in the TimerFuture's code, but
+        // in the wild we won't have control over how the waker is called.
+        // (That state locking issue isn't something I made up or introduced on
+        // purpose, it's genuinely the code from the async book.)
+    }
+}
+
+#[pymethods]
+impl WakerClosure {
+    // I didn't change anything here, but as an exercise, try and move the code
+    // from this function to `wake_by_ref` above to skip the double callback,
+    // and see what happens.
```

Comment:
In fact I did try to merge the two together myself and could not reproduce a deadlock...

Maybe it's because I'm not using any Mutex in the waker this time around, or maybe it's because this too contrived an example with simple deterministicly waiting threads.

I hope I'm wrong in what I wrote here, and that there is no need for the double callback trick in the end, but I don't hold too much hope, so I leave that here for the time being.

I'll be doing some more testing with real-world examples to be sure.